### PR TITLE
Feature card supertype screen

### PR DESCRIPTION
**Pull request by MarcosCGlobant**
PR created to merge the Pokemon type fragment feature with master.

Tests of the feature with coverage:

tests passed pokemon supertype feature

supertype viewmodel tests coverage

Stetho debugging the database:

stetho database debugg

App GIF:

pokemon supertypes database

### CODE REVIEW

**Review by lucianoeli — APPROVED** [review]

File: data/src/main/java/com/globant/data/service/response/PokemonSupertypesResponse.kt [hunk 1]
```diff
@@ -0,0 +1,5 @@
+package com.globant.data.service.response
+
+class PokemonSupertypesResponse<T>(
```
Comment (SartiMau) [inline data/src/main/java/com/globant/data/service/response/PokemonSupertypesResponse.kt]: Why do you need a generic here? The response isn't String?

File: data/src/main/java/com/globant/data/service/response/PokemonTypesResponse.kt [hunk 2]
```diff
@@ -1,5 +1,5 @@
 package com.globant.data.service.response
 
-class PokemonTypesBaseResponse<T>(
+class PokemonTypesResponse<T>(
```
Comment (SartiMau) [inline data/src/main/java/com/globant/data/service/response/PokemonTypesResponse.kt]: same

File: data/src/main/java/com/globant/data/service/response/PokemonSupertypesResponse.kt (hunk 1 above)
Comment (MarcosCGlobant) [reply data/src/main/java/com/globant/data/service/response/PokemonSupertypesResponse.kt]: that's true, changed the generic for the actual response type

File: data/src/main/java/com/globant/data/service/response/PokemonTypesResponse.kt (hunk 2 above)
Comment (MarcosCGlobant) [reply data/src/main/java/com/globant/data/service/response/PokemonTypesResponse.kt]: good catch, thanks

**Review by SartiMau — APPROVED** [review]

File: app/src/main/res/layout/fragment_pokemon_supertype_layout.xml [hunk 3]
```diff
@@ -1,15 +1,32 @@
...
+        android:orientation="vertical"
+        android:visibility="gone">
+
+        <ProgressBar
+            android:id="@+id/pokemon_supertype_progress_bar"
+            style="?android:attr/progressBarStyle"
+            android:layout_width="wrap_content"
+            android:layout_height="wrap_content" />
+
+    </LinearLayout>
```
Comment (nicolbidone) [inline app/src/main/res/layout/fragment_pokemon_supertype_layout.xml]: why the progress bar is inside a linear layout?

File: domain/src/main/java/com/globant/domain/service/PokemonSupertypesService.kt [hunk 4]
```diff
@@ -0,0 +1,8 @@
+package com.globant.domain.service
+
+import com.globant.domain.entity.PokemonSupertype
+import com.globant.domain.util.Result
+
+interface PokemonSupertypesService {
+    fun getPokemonSupertypesFromAPI(pokemonSupertypesResources: MutableMap<String, Int>): Result<List<PokemonSupertype>>
```
Comment (nicolbidone) [inline domain/src/main/java/com/globant/domain/service/PokemonSupertypesService.kt]: it is necessary to have the suffix FromAPI?

**Review by mauroLonghi — APPROVED** [review]

File: data/src/main/java/com/globant/data/database/entity/PokemonSupertypeRoom.kt [hunk 5]
```diff
@@ -0,0 +1,10 @@
+package com.globant.data.database.entity
+
+import androidx.room.Entity
+import androidx.room.PrimaryKey
+
+@Entity(tableName = "pokemon_supertype_table")
+data class PokemonSupertypeRoom(
```
Comment (nicolbidone) [inline data/src/main/java/com/globant/data/database/entity/PokemonSupertypeRoom.kt]: this class should be PokemonSupertypeEntity, if you change the database you should rename everything

File: data/src/main/java/com/globant/data/mapper/PokemonSupertypeRoomMapper.kt [hunk 6]
```diff
@@ -0,0 +1,21 @@
+package com.globant.data.mapper
+
+import com.globant.data.database.entity.PokemonSupertypeRoom
+import com.globant.domain.entity.PokemonSupertype
+
+class PokemonSupertypeRoomMapper : BaseMapper<List<PokemonSupertypeRoom>, MutableList<PokemonSupertype>, MutableMap<String, Int>?> {
```
Comment (nicolbidone) [inline data/src/main/java/com/globant/data/mapper/PokemonSupertypeRoomMapper.kt]: this should be PokemonSupertypeDataBaseMapper

File: data/src/main/java/com/globant/data/mapper/PokemonTypeRoomMapper.kt [hunk 7]
```diff
@@ -3,8 +3,8 @@ package com.globant.data.mapper
 import com.globant.data.database.entity.PokemonTypeRoom
 import com.globant.domain.entity.PokemonType
 
-class PokemonTypeRoomMapper : BaseMapper<List<PokemonTypeRoom>, MutableList<PokemonType>> {
-    override fun transform(type: List<PokemonTypeRoom>, resourcesMap: MutableMap<String, Pair<Int, Int>>?): MutableList<PokemonType> {
+class PokemonTypeRoomMapper : BaseMapper<List<PokemonTypeRoom>, MutableList<PokemonType>, MutableMap<String, Pair<Int, Int>>?> {
```
Comment (nicolbidone) [inline data/src/main/java/com/globant/data/mapper/PokemonTypeRoomMapper.kt]: same

File: app/src/main/res/layout/fragment_pokemon_supertype_layout.xml (hunk 3 above)
Comment (MarcosCGlobant) [reply app/src/main/res/layout/fragment_pokemon_supertype_layout.xml]: To make the loader more visible, changed to Relative Layout to make it less consuming, thanks for the help

File: domain/src/main/java/com/globant/domain/service/PokemonSupertypesService.kt (hunk 4 above)
Comment (MarcosCGlobant) [reply domain/src/main/java/com/globant/domain/service/PokemonSupertypesService.kt]: it's not necessary, refactored, thanks

File: data/src/main/java/com/globant/data/database/entity/PokemonSupertypeRoom.kt (hunk 5 above)
Comment (MarcosCGlobant) [reply data/src/main/java/com/globant/data/database/entity/PokemonSupertypeRoom.kt]: refactored, thanks

File: data/src/main/java/com/globant/data/mapper/PokemonSupertypeRoomMapper.kt (hunk 6 above)
Comment (MarcosCGlobant) [reply data/src/main/java/com/globant/data/mapper/PokemonSupertypeRoomMapper.kt]: refactored, thanks

File: data/src/main/java/com/globant/data/mapper/PokemonTypeRoomMapper.kt (hunk 7 above)
Comment (MarcosCGlobant) [reply data/src/main/java/com/globant/data/mapper/PokemonTypeRoomMapper.kt]: refactored, thanks

File: data/src/main/java/com/globant/data/database/PokemonSupertypeDatabaseImpl.kt [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+
+import com.globant.data.mapper.PokemonSupertypeDatabaseMapper
+import com.globant.domain.database.PokemonSupertypeDatabase
+import com.globant.domain.entity.PokemonSupertype
+import com.globant.domain.util.Constant
+import com.globant.domain.util.Result
+
+class PokemonSupertypeDatabaseImpl(private val pokemonDao: PokemonDao) : PokemonSupertypeDatabase {
+
+    private val pokemonSupertypeRoomMapper = PokemonSupertypeDatabaseMapper()
```
Comment (nicolbidone) [inline data/src/main/java/com/globant/data/database/PokemonSupertypeDatabaseImpl.kt]: Missed Room Reference everywhere

Comment (MarcosCGlobant) [reply data/src/main/java/com/globant/data/database/PokemonSupertypeDatabaseImpl.kt]: Removed all the references, thanks

File: data/src/main/java/com/globant/data/database/PokemonDao.kt [hunk 9]
```diff
@@ -4,13 +4,20 @@ import androidx.room.Dao
...
+import com.globant.data.database.entity.PokemonSupertypeDatabaseEntity
+import com.globant.data.database.entity.PokemonTypeDatabaseEntity
 
 @Dao
 interface PokemonDao {
     @Query("SELECT * FROM pokemon_type_table")
-    fun getPokemonTypes(): List<PokemonTypeRoom>
+    fun getPokemonTypes(): List<PokemonTypeDatabaseEntity>
 
     @Insert(onConflict = OnConflictStrategy.IGNORE)
```
Comment (SartiMau) [inline data/src/main/java/com/globant/data/database/PokemonDao.kt]: this strategy is not the correct, should be REPLACE. In this way, you will always save the last info

File: data/src/main/java/com/globant/data/database/PokemonDao.kt [hunk 10]
```diff
@@ -4,13 +4,20 @@ import androidx.room.Dao
...
-    fun insertPokemonType(pokemonTypeRoom: PokemonTypeRoom)
+    fun insertPokemonType(pokemonTypeDatabaseEntity: PokemonTypeDatabaseEntity)
+
+    @Query("SELECT * FROM pokemon_supertype_table")
+    fun getPokemonSupertypes(): List<PokemonSupertypeDatabaseEntity>
+
+    @Insert(onConflict = OnConflictStrategy.IGNORE)
```
Comment (SartiMau) [inline data/src/main/java/com/globant/data/database/PokemonDao.kt]: same

**Review by SartiMau — CHANGES_REQUESTED** [review]
see comments pls

File: data/src/main/java/com/globant/data/database/PokemonDao.kt (hunk 9 above)
Comment (MarcosCGlobant) [reply data/src/main/java/com/globant/data/database/PokemonDao.kt]: refactored, thanks

File: data/src/main/java/com/globant/data/database/PokemonDao.kt (hunk 10 above)
Comment (MarcosCGlobant) [reply data/src/main/java/com/globant/data/database/PokemonDao.kt]: refactored, thanks

**Review by SartiMau — APPROVED** [review]

**Review by nicolbidone — APPROVED** [review]

**Pull request merged**